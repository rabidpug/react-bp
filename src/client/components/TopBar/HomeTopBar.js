// @flow

import { getIsOnline, getIsSidebarCollapsed, } from 'Store/ui/selectors';

import SlideWrap from 'Animations/SlideWrap';
import TopBar from 'Components/TopBar';
import { connect, } from 'react-redux';
import { getIsAuthenticated, } from 'Store/user/selectors';
import { toggleSidebar, } from 'Store/ui/actions';

const mapStateToProps = (
  state, ownProps
) => ( {
  actionMenuItems    : ownProps.route.actionMenuItems,
  isAuthenticated    : getIsAuthenticated( state ),
  isOnline           : getIsOnline( state ),
  isSidebarCollapsed : getIsSidebarCollapsed( state ),
  selectedActionKeys : [],
} );

const mapDispatchToProps = dispatch => ( { toggleSideBar: () => dispatch( toggleSidebar() ), } );

const HomeTopBar = connect(
  mapStateToProps, mapDispatchToProps
)( SlideWrap( TopBar ) );

export default HomeTopBar;
