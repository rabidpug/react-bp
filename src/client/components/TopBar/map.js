import { getIsOnline, getIsSidebarCollapsed, } from 'Store/ui/selectors';

import { getIsAuthenticated, } from 'Store/user/selectors';
import { toggleSidebar, } from 'Store/ui/actions';

const mapTopBar = {
  Dispatch : dispatch => ( { toggleSideBar: () => dispatch( toggleSidebar() ), } ),
  State    : ( state, ownProps ) => ( {
    actionMenuItems    : ownProps.route.actionMenuItems,
    isAuthenticated    : getIsAuthenticated( state ),
    isOnline           : getIsOnline( state ),
    isSidebarCollapsed : getIsSidebarCollapsed( state ),
    selectedActionKeys : [],
  } ),
};

export default mapTopBar;
