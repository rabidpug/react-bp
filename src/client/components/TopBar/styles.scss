.trigger {
  font-size: 18px;
  line-height: 64px;
  padding: 0 24px;
  cursor: pointer;
  transition: color 0.3s;
  float: left;
}
.trigger:hover {
  color: #1890ff;
}

.topBarMenu {
  float: right;
}
