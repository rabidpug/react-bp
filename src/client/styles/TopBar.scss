@import 'theme';

.topBarMenu {
  background-color: rgba(0, 0, 0, 0);
  float: right;
}

.topBarHeader {
  background: rgb(255, 255, 255);
  border-top: 0.25rem solid $primary-color;
  box-shadow: 0 0.0625rem 0.375rem 0 rgba(0, 0, 0, 0.2);
  padding: 0;
  z-index: 1;
}

.trigger {
  background-color: rgba(0, 0, 0, 0);
  margin-left: 0.625rem;
}

.offlineSpinContainer {
  float: right;
  margin: 0 0.625rem;
}

.offlineSpinner {
  color: #737373;
}

.offlineSpinner i {
  background-color: #f5222d;
}
