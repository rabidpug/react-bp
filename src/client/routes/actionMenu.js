import store, { reduxStore, } from 'Store';

import TopBar from 'Components/TopBar';

const actionMenu = [
  {
    actionMenuItems: [
      {
        action          : () => reduxStore.dispatch( store.user.set.logoutUser() ),
        icon            : 'logout',
        isAuthenticated : true,
        key             : '1',
        label           : 'Log Out',
      },
    ],
    component : TopBar,
    path      : '/profile',
  },
  {
    actionMenuItems : [],
    component       : TopBar,
    path            : '*',
  },
];

export default actionMenu;
