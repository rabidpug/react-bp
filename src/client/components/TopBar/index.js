import {
  Icon,
  Layout,
  Menu,
} from 'antd';

import React from 'react';
import styles from './styles.scss';

const { Item, } = Menu;

const { Header, } = Layout;
const TopBar = ( {
  isSidebarCollapsed, toggleSideBar, actionMenuItems,
} ) => (
  <Header
    style={ { background : '#fff',
              padding    : 0, } }>
    <Icon
      className={ styles.trigger }
      onClick={ toggleSideBar }
      type={ isSidebarCollapsed ? 'menu-unfold' : 'menu-fold' } />
    {actionMenuItems && (
      <Menu
        className={ styles.topBarMenu }
        mode='horizontal'
        onClick={ ( { item, } ) => item.props.action() }
        selectedKeys={ [] }>
        {actionMenuItems.map( item => (
          <Item
            action={ item.action }
            key={ item.key }>
            <Icon type={ item.icon } />
            {item.label}
          </Item>
        ) )}
      </Menu>
    )}
  </Header>
);

export default TopBar;
