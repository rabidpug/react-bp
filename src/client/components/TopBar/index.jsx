import { Button, Icon, Layout, Menu, Spin, } from 'antd';
import React, { Component, } from 'react';

import SlideWrap from 'Animations/SlideWrap';
import { connect, } from 'react-redux';
import mapTopBar from './map';
import styles from 'Styles/TopBar';

const { Item, SubMenu, } = Menu;

const { Header, } = Layout;

@connect( mapTopBar.State, mapTopBar.Dispatch )
@SlideWrap
export default class TopBar extends Component {
  render () {
    const {
      isSidebarCollapsed,
      toggleSideBar,
      actionMenuItems,
      selectedActionKeys,
      isAuthenticated,
      style,
      isOnline,
    } = this.props;
    const filterAuth = item => typeof item.isAuthenticated !== 'boolean' || item.isAuthenticated === isAuthenticated;

    const menuItemMap = item =>
      item.subMenu ? (
        <SubMenu
          key={ item.key }
          title={
            <span>
              <Icon type={ item.icon } />
              <span>{item.label}</span>
            </span>
          }>
          {item.subMenu.filter( filterAuth ).map( menuItemMap )}
        </SubMenu>
      ) : (
        <Item
          action={ item.action }
          key={ item.key }>
          {item.component
            ? <item.component { ...item } />
            : (
              <span>
                <Icon type={ item.icon } />
                {item.label}
              </span>
            )}
        </Item>
      );
    const onlineStatusStyles = { background: isOnline ? '#fff' : 'rgba(117,117,117,0.4)', };

    if ( !isOnline ) onlineStatusStyles.borderTop = '4px solid #f5222d';

    return (
      <Header
        className={ styles.topBarHeader }
        style={ {
          ...onlineStatusStyles,
          padding: 0,
        } }>
        <Button
          className={ styles.trigger }
          icon={ isSidebarCollapsed ? 'menu-unfold' : 'menu-fold' }
          onClick={ toggleSideBar }
        />
        {!isOnline && (
          <div className={ styles.offlineSpinContainer }>
            <Spin
              className={ styles.offlineSpinner }
              tip='Offline...' />
          </div>
        )}
        {actionMenuItems && (
          <Menu
            className={ styles.topBarMenu }
            mode='horizontal'
            onClick={ ( { item, } ) => item.props.action && item.props.action() }
            selectedKeys={ typeof selectedActionKeys === 'object' ? selectedActionKeys : [ selectedActionKeys, ] }
            style={ { ...style, } }>
            {actionMenuItems.filter( filterAuth ).map( menuItemMap )}
          </Menu>
        )}
      </Header>
    );
  }
}
