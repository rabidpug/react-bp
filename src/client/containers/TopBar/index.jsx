import React, { Component, } from 'react';
import { faChevronLeft, faChevronRight, } from '@fortawesome/free-solid-svg-icons';

import Button from 'Components/Button';
import { FontAwesomeIcon, } from '@fortawesome/react-fontawesome';
import Layout from 'Components/Layout';
import Loading from 'Components/Loading';
import Menu from 'Components/Menu';
import { connect, } from 'react-redux';
import mapTopBar from './map';
import mappedMenus from 'Helpers/mappedMenus';
import styles from 'Styles/TopBar';

@connect( mapTopBar.State, mapTopBar.Dispatch )
export default class TopBar extends Component {
  constructor ( props ) {
    super( props );

    this.state = { openKeys: [], };
  }

  toggleSubMenu = key => {
    this.setState( prevState => {
      const openKeys = [ ...prevState.openKeys, ];

      openKeys.includes( key ) ? openKeys.splice( openKeys.indexOf( key ), 1 ) : openKeys.push( key );

      return {
        ...prevState,
        openKeys,
      };
    } );
  };

  render () {
    const {
      isSidebarCollapsed,
      toggleSidebar,
      actionMenuItems,
      selectedActionKeys,
      isAuthenticated,
      isOnline,
    } = this.props;
    const { openKeys, } = this.state;
    const onlineStatusStyles = { background: isOnline ? '#fff' : 'rgba(117,117,117,0.4)', };

    if ( !isOnline ) onlineStatusStyles.borderTop = '4px solid #f5222d';

    return (
      <Layout.Header>
        <Button
          onClick={ () => toggleSidebar() }
          style={ {
            float  : 'left',
            height : '2.7rem',
          } }
          variant='secondary'>
          <FontAwesomeIcon icon={ isSidebarCollapsed ? faChevronRight : faChevronLeft } />
        </Button>
        {!isOnline && (
          <div className={ styles.offlineSpinContainer }>
            <Loading
              className={ styles.offlineSpinner }
              tip='Offline...' />
          </div>
        )}
        {actionMenuItems && (
          <Menu
            openKeys={ openKeys }
            selectedKeys={ typeof selectedActionKeys === 'object' ? selectedActionKeys : [ selectedActionKeys, ] }
            showText
            subMenuAction={ this.toggleSubMenu }>
            {mappedMenus( actionMenuItems, isAuthenticated )}
          </Menu>
        )}
      </Layout.Header>
    );
  }
}
