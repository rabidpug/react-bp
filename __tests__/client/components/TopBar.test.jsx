import React from 'react';
import TopBar from 'Components/TopBar';
import actionMenu from 'Routes/actionMenu';
import { shallow, } from 'enzyme';
import { shallowToJson, } from 'enzyme-to-json';

describe(
  'TopBar', () => {
    const [ { actionMenuItems, }, ] = actionMenu.filter( menu => menu.path === '*' );
    const wrapper = shallow( <TopBar actionMenuItems={ actionMenuItems } /> );

    it(
      'Should render the TopBar', () => {
        expect( shallowToJson( wrapper ) ).toMatchSnapshot();
      }
    );

    it(
      'Should perform the action of the menu item clicked', () => {
        const action = jest.fn();

        wrapper.find( 'Menu' ).simulate(
          'click', { item: { props: { action, }, }, }
        );

        expect( action ).toBeCalled();
      }
    );
  }
);
