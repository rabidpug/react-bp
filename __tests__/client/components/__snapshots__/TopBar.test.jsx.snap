// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopBar Should render the TopBar 1`] = `
<Adapter
  className="topBarHeader"
  style={
    Object {
      "background": "rgba(117,117,117,0.4)",
      "borderTop": "4px solid #f5222d",
      "padding": 0,
    }
  }
>
  <Button
    className="trigger"
    ghost={false}
    icon="menu-fold"
    loading={false}
    prefixCls="ant-btn"
  />
  <div
    className="offlineSpinContainer"
  >
    <Spin
      className="offlineSpinner"
      prefixCls="ant-spin"
      size="default"
      spinning={true}
      tip="Offline..."
      wrapperClassName=""
    />
  </div>
  <Menu
    className="topBarMenu"
    mode="horizontal"
    onClick={[Function]}
    prefixCls="ant-menu"
    selectedKeys={
      Array [
        undefined,
      ]
    }
    style={Object {}}
    theme="light"
  />
</Adapter>
`;
